src/haxeLanguageServer/server/HaxeServer.hx

```diff
@@ -186,21 +166,6 @@ class HaxeServer {
 		trace(message + "...");
 	}
 
-	function hasNonCancellableRequests():Bool {
```

Comment:
Speaking of removed things, what happened to this?

Reply:
Yes I wanted to ask about that because haxe.exe doesn't (and won't) have the concept of cancellation. So I'm not sure how exactly this comes into play here...

Reply:
"Cancel" just refers to server restarts in this case. We don't want to restart a server (for instance due to settings changes) while there's a compilation going on.

Reply:
In that case something doesn't look quite right on master already because this is only checked on `start`, while `stop` is also a public function. I don't see anything preventing a `proc.kill()` while `hasNonCancellableRequests()` returns `true`.

Reply:
Isn't `stop()` only called on shutdown requests?

Reply:
I'm also seeing it in `Context.onExit`, which suggests to me that simply closing the language server could abort a compilation mid-process.

Reply:
Well, I think that's expeced? Or would you want compilation to continue in the background after closing VSCode? :)

Reply:
I just don't understand why we would care about this when restarting, but not when stopping. If bad things can happen then we shouldn't just kill the process either way, at least not on a normal exit. And if no bad things can happen then why do we have to wait on a restart?

Reply:
A server restart doesn't qualify as "bad things" though, it happens all the time. Imagine changing the target you want code completion for during a compilation that's taking a while - killing the server in the middle of that is highly unexpected behavior.

Reply:
> it happens all the time

But why? The whole idea of a server is to keep it running, not restart it all the time.

Anyway, if we need this then we have to bring back a queue for the display requests. In that case we would actually end up with _two_ queues because the process already manages one. I guess that's just necessary though...

Reply:
Yeah, perhaps we should do it all the time. But that's just how it works atm, stuff like cache builds and reading classpaths for instance only happens when the server is started up, we would need some way to do that when switching targets without restarts. I'm also somewhat concerned about server stability / memory usage. If you have a big project and are keeping everything the results of 3 different compilation targets in memory all the time, that might be bad? Not sure.

Reply:
Related: https://github.com/vshaxe/vshaxe/issues/218

Reply:
I looked at this again with less distractions and realized what we're actually after here is just a way to gracefully stop the process. I'm gonna change the `close` function of the processes to this:

```
function close(graceful = true, callback:() -> Void)
```

If `graceful == true` and there's an ongoing request, so let that request finish and then call `callback`. If there is no request, we call `callback` immediately.

If `graceful == false` we just kill process and call `callback`.

This allows us to implement restart by doing something like `process.close(true, start)`.

Reply:
Hm... this only really matters for actual compilation requests though. It could actually be problematic to do it for others, for instance if you switch to another target and back immediately - you'd have to go through a lot of pointless cache builds if they all just queue up.

Reply:
Well, then we really have to distinguish this stuff at vshaxe level I suppose. I didn't realize it required intervention at such a low level. In that case it is probably going to require its own process handling and trying to use haxeserver is a pointless endeavor.

Reply:
Just for this one thing? Can't there simply be some "cancelable" flag when starting a request?

Reply:
That is not a property of requests as far as Haxe is concerned. You want to distinguish whether or not you _should_ cancel a request. That is something that you as the client of haxeserver should manage, and then set the `graceful` argument accordingly. Otherwise we're mixing abstractions again already.

And if we have to manage a queue for requests at vshaxe-level anyway I'm not sure it makes sense to use haxeserver. We can however try if it works.

Reply:
I shouldn't use the term "client" when servers and clients are actually involved, think "user" instead.

Reply:
Well, what we really want to know is whether the request that's currently in progress can be cancelled. haxeserver would just need to expose some way to check what the current request is? Of course then there needs to be some way to identify a request, maybe `rawRequest` should return an incremeting ID or something..

Reply:
That might be a good idea. It would allow users to associate any kind of information with a given request while haxeserver remains oblivious.

Reply:
Hm, I guess there could be a cancelable request in progress while a non-cancelable is queued up. But haxeserver could still expose the current request view to be inspected somehow.

Reply:
To haxeserver, a request is just a byte sequence that is sent to the haxe process.

I'm a bit concerned that this is starting to sound even messier than the current implementation...

Reply:
Actually, there's callbacks, so haxeserver might not need to expose a queue at all. haxe-languageserver could maybe simply increment a counter that keeps track of unfinished non-cancelable requests when calling `rawRequest`, and decrement when a callback of one is invoked.

Reply:
That's basically what I meant by "bring back a queue for the display requests". There's not that much difference between incrementing/decrementing a counter and adding to/removing from a list.

Reply:
Ah, to me it sounded like you want to have haxe-language-server manage the execution of the requests as well / not really use haxe-server at all.

Reply:
Well, I'm looking at it and see 2 queues, and wonder if that's really better than 1 queue. So I'm going back and forth between all this being a good or bad idea.

Reply:
I don't think having some additional list or counter really impacts the main motivation behind this (avoiding code duplication between LS and Haxe tests). haxe-server still handles all the low-level details.

Reply:
Yes but then I worry that the LS actually cares about the low-level details. There's also this entire `onExit` business which randomly inspects `buffer` to determine some error causes.

Reply:
Maybe haxe-server could provide an abstraction for that, like an `onExit(reason:InvalidArgument|Shutdown|Unknown)` event you can add a listener for.

Reply:
I have added `onExit`, but the callback only receives a string. I don't want to add any half-assed message parsing to haxe-server. Something like this should probably be solved by Haxe using defined exit codes.